File: decl-py/src/decl/qengine/edits.py
"""Replay Session root binding over retained records and member revisions."""

from __future__ import annotations

from collections.abc import Callable
from copy import copy
from typing import TYPE_CHECKING, Any

from decl.qengine.revisions import Revisions
from decl.semantics import (
    ABSENT,
    ArrV,
    JObj,
    MapV,
    PreVal,
    Quantity,
    RangeV,
    RecInst,
    Ref,
    path_str,
    value_eq,
)

if TYPE_CHECKING:
    from decl.engine import Engine


def same_raw(a: Any, b: Any) -> bool:
    if a is b:
        return True
    if type(a) is not type(b):
        return False
    if isinstance(a, list):
        return len(a) == len(b) and all(same_raw(x, y) for x, y in zip(a, b, strict=True))
    if isinstance(a, JObj):
        return len(a.entries) == len(b.entries) and all(
            k == l and same_raw(x, y) for (k, x), (l, y) in zip(a.entries, b.entries, strict=True)
        )
    if isinstance(a, PreVal):
        x, y = a.scope, b.scope
        return (
            a.expr is b.expr
            and x.inst is y.inst
            and x.menv is y.menv
            and x.root_name == y.root_name
            and x.locals.keys() == y.locals.keys()
            and all(same_raw(v, y.locals[k]) for k, v in x.locals.items())
        )
    if a is None or type(a) in (int, float, str, bool) or isinstance(a, Quantity):
        return value_eq(a, b)
    return False


def capture(v: Any) -> Callable[[Any], bool]:
    if isinstance(v, RecInst):
        rt, owner, path, order = v.rt, v.eng, list(v.path), list(v.entry_order)
        members = [(n, s.kind, s.hidden) for n, s in v.slots.items()]
        extras = [(k, capture(x)) for k, x in v.extras.items()]
        return lambda n: (
            isinstance(n, RecInst)
            and n is v
            and n.rt is rt
            and n.eng is owner
            and n.path == path
            and n.entry_order == order
            and [(k, s.kind, s.hidden) for k, s in n.slots.items()] == members
            and list(n.extras) == [k for k, _ in extras]
            and all(matches(n.extras[k]) for k, matches in extras)
        )
    if isinstance(v, ArrV):
        path, items = list(v.path), [capture(x) for x in v.items]
        return lambda n: (
            isinstance(n, ArrV)
            and n.path == path
            and len(n.items) == len(items)
            and all(f(x) for f, x in zip(items, n.items, strict=True))
        )
    if isinstance(v, MapV):
        path, entries = list(v.path), [(k, capture(x)) for k, x in v.entries.items()]
        return lambda n: (
            isinstance(n, MapV)
            and n.path == path
            and list(n.entries) == [k for k, _ in entries]
            and all(f(n.entries[k]) for k, f in entries)
        )
    if isinstance(v, Ref):
        if v.snap or v.inverse or v.round_ref:
            return lambda _: False
        return lambda n: (
            isinstance(n, Ref) and not (n.snap or n.inverse or n.round_ref) and value_eq(v, n)
        )
    if (
        v is None
        or v is ABSENT
        or type(v) in (int, float, str, bool)
        or isinstance(v, (Quantity, RangeV))
    ):
        return lambda n: value_eq(v, n)
    return lambda _: False


class Edits:
    def __init__(self) -> None:
        self.active = False
        self.revisions = Revisions()
        self.slot_computes = 0
        self.retained_records = 0
        self.prepared_queries = 0
        self.inputs: dict[RecInst, dict[str, Any]] = {}
        self.roots: dict[str, tuple[Any, bool]] = {}
        self.pool: dict[str, RecInst] = {}
        self.live: set[RecInst] = set()
        self.invalid: set[str] = set()
        self.root_values: dict[str, Any] = {}
        self.root_reads: dict[str, set[str]] = {}
        self.binding_reuse = True
        self.revisions.resolve = self.resolve

    @staticmethod
    def resolve(eng: Engine, key: str) -> bool:
        slot = eng.slots_by_key.get(key)
        if slot:
            eng.force_slot(*slot)
            return True
        if key.startswith("const:"):
            index, name = key[6:].split("|", 1)
            envs = list(eng.const_envs)
            if int(index) < len(envs) and name in envs[int(index)].consts:
                eng.force_const_in(envs[int(index)], name, "")
                return True
        return False

    def retain(self, eng: Engine, v: Any, seen: set[int]) -> bool:
        if (
            v is None
            or v is ABSENT
            or type(v) in (int, float, str, bool)
            or isinstance(v, (Quantity, RangeV))
        ):
            return True
        if id(v) in seen:
            return True
        seen.add(id(v))
        if isinstance(v, Ref):
            return not (v.snap or v.inverse or v.round_ref)
        if isinstance(v, RecInst):
            return v.eng is eng and all(
                self.retain(eng, x, seen) for x in self.inputs.get(v, {}).values()
            )
        if isinstance(v, ArrV):
            return all(self.retain(eng, x, seen) for x in v.items)
        if isinstance(v, MapV):
            return all(self.retain(eng, x, seen) for x in v.entries.values())
        if isinstance(v, list):
            return all(self.retain(eng, x, seen) for x in v)
        if isinstance(v, JObj):
            return all(self.retain(eng, x, seen) for _, x in v.entries)
        if isinstance(v, PreVal):
            return (v.scope.inst is None or v.scope.inst.eng is eng) and all(
                self.retain(eng, x, seen) for x in v.scope.locals.values()
            )
        return False

    def begin(self, eng: Engine, changed: set[str], documents: dict[str, Any]) -> None:
        values: dict[str, Any] = {}
        forced: set[str] = set()
        errors = bool(eng.env.diagnostics)
        self.binding_reuse = not errors
        self.pool = {path_str(r.path): r for r in eng.env.registry}
        self.live.clear()
        self.retained_records = 0
        self.root_values = dict(eng.env.roots)
        self.root_reads = {f"root:{n}": set(eng.reads.get(f"root:{n}", ())) for n in eng.env.roots}
        known_inputs: set[str] = set()

        def collect(v: Any, into: set[str]) -> None:
            if isinstance(v, RecInst):
                for n, slot in v.slots.items():
                    key = f"{path_str(v.path)}.{n}"
                    eng.slots_by_key[key] = (v, n)
                    into.add(key)
                    if slot.state == "ok":
                        collect(slot.value, into)
            elif isinstance(v, ArrV):
                for x in v.items:
                    collect(x, into)
            elif isinstance(v, MapV):
                for x in v.entries.values():
                    collect(x, into)

        def diff(before: Any, next_: Any, value: Any, producer: str) -> None:
            if same_raw(before, next_):
                return
            known_inputs.add(producer)
            forced.add(producer)
            if isinstance(value, RecInst) and isinstance(before, JObj) and isinstance(next_, JObj):
                a, b = dict(before.entries), dict(next_.entries)
                for name in a.keys() | b.keys():
                    if (name in a) == (name in b) and same_raw(a.get(name), b.get(name)):
                        continue
                    slot = value.slots.get(name)
                    if slot:
                        key = f"{path_str(value.path)}.{name}"
                        eng.slots_by_key[key] = (value, name)
                        diff(a.get(name), b.get(name), slot.value, key)
                        forced.add(key)
            elif isinstance(value, ArrV) and isinstance(before, list) and isinstance(next_, list):
                for i in range(max(len(before), len(next_))):
                    diff(
                        before[i] if i < len(before) else None,
                        next_[i] if i < len(next_) else None,
                        value.items[i] if i < len(value.items) else None,
                        producer,
                    )
            elif isinstance(value, MapV) and isinstance(before, JObj) and isinstance(next_, JObj):
                a, b = dict(before.entries), dict(next_.entries)
                for k in a.keys() | b.keys():
                    diff(a.get(k), b.get(k), value.entries.get(k), producer)
            else:
                collect(value, forced)

        for name in changed:
            previous, value, key = self.roots.get(name), eng.env.roots.get(name), f"root:{name}"
            forced.add(key)
            known_inputs.add(key)
            if previous is not None and not previous[1] and name in documents:
                diff(previous[0], documents[name], value, key)
            else:
                collect(value, forced)
        readers: dict[str, list[str]] = {}
        for key, deps in eng.reads.items():
            for dep in deps:
                readers.setdefault(dep, []).append(key)
                if dep.startswith(("edge:", "snapshot:", "round:", "value:")):
                    forced.add(key)
        for name, value in eng.env.roots.items():
            values[f"root:{name}"] = value
        for inst in eng.env.registry:
            for name, slot in inst.slots.items():
                key = f"{path_str(inst.path)}.{name}"
                if slot.state in ("ok", "absent"):
                    values[key] = ABSENT if slot.state == "absent" else slot.value
                if errors:
                    forced.add(key)
                    eng.slots_by_key[key] = (inst, name)
        for index, env in enumerate(eng.const_envs):
            for name, con in env.consts.items():
                key = f"const:{index}|{name}"
                if con.get("state") == "ok":
                    values[key] = con.get("value")
                if errors:
                    forced.add(key)
        if eng.queried:
            for key, value in values.items():
                if not self.retain(eng, value, set()):
                    forced.add(key)
        if errors:
            forced.update(eng.reads)
        invalid: set[str] = set()
        queue = list(forced)
        if any(d.startswith("value:") for deps in eng.reads.values() for d in deps):
            queue.extend(values)
        while queue:
            key = queue.pop()
            if key in invalid:
                continue
            invalid.add(key)
            if key not in known_inputs and key in values:
                descendants: set[str] = set()
                collect(values[key], descendants)
                queue.extend(descendants - invalid)
            queue.extend(readers.get(key, ()))
        self.invalid = invalid
        self.prepared_queries = len(invalid)
        self.revisions.begin_queries(eng, invalid, values, forced, capture)
        copied: set[RecInst] = set()
        for key in invalid:
            found = eng.slots_by_key.get(key)
            if found is None:
                continue
            inst, name = found
            slot = inst.slots.get(name)
            if slot is None or slot.compute is None:
                continue
            if inst not in copied:
                inst.slots = dict(inst.slots)
                copied.add(inst)
            next_slot = copy(slot)
            next_slot.state, next_slot.value = "unforced", None
            inst.slots[name] = next_slot
        for index, env in enumerate(eng.const_envs):
            for name, con in env.consts.items():
                if f"const:{index}|{name}" in invalid:
                    con["state"] = "unforced"
                    con.pop("value", None)
        eng.env.roots.clear()
        eng.env.registry.clear()
        eng.env.diagnostics.clear()
        eng.failed_inputs.clear()
        eng.deferred_slots = []
        eng.deferred_roots = []
        eng.phase, eng.settled, eng.prev, eng.snap = 1, False, None, None
        eng.queried.clear()
        eng.ref_index.clear()
        eng.computing_edges.clear()
        eng.edge_bases = []
        eng.round_roots, eng.snapshot_value = None, None
        self.active = True

    def root(
        self, eng: Engine, name: str, raw: Any, expression: bool, run: Callable[[], Any]
    ) -> Any:
        previous = self.roots.get(name)
        self.roots[name] = (raw, expression)
        if not self.active:
            return run()
        key = f"root:{name}"
        if (
            previous is None
            or previous[1] != expression
            or not (previous[0] is raw if expression else same_raw(previous[0], raw))
        ):
            self.revisions.force(key)
        if key not in self.invalid and name in self.root_values:
            eng.reads[key] = self.root_reads.get(key, set())
            value = self.root_values[name]
            self.activate(eng, value)
            return value
        return self.compute(eng, key, run)

    def unchanged(self, inst: RecInst, entries: Any) -> bool:
        before = self.inputs.get(inst)
        return (
            self.binding_reuse
            and before is not None
            and len(before) == len(entries)
            and inst.entry_order == [k for k, _ in entries]
            and all(k in before and same_raw(before[k], v) for k, v in entries)
        )

    def record(self, rt: Any, path: Any, parent: RecInst | None) -> RecInst | None:
        if not self.active:
            return None
        old = self.pool.get(path_str(path))
        return old if old is not None and old.rt is rt and old.parent is parent else None

    def bound(self, eng: Engine, inst: RecInst, entries: Any, old_slots: Any = None) -> None:
        before, next_ = self.inputs.get(inst), dict(entries)
        if self.active:
            for name, slot in inst.slots.items():
                key = f"{path_str(inst.path)}.{name}"
                old = old_slots.get(name) if old_slots is not None else None
                if (
                    before is None
                    or (name in before) != (name in next_)
                    or not same_raw(before.get(name), next_.get(name))
                    or old is None
                    or old.kind != slot.kind
                ):
                    self.revisions.force(key)
                elif old is not None:
                    inst.slots[name] = old
                eng.slots_by_key[key] = (inst, name)
        self.inputs[inst] = next_
        if self.active:
            for slot in inst.slots.values():
                if slot.state == "ok":
                    self.activate(eng, slot.value)

    def register(self, eng: Engine, inst: RecInst) -> None:
        if inst in self.live:
            return
        self.live.add(inst)
        eng.env.registry.append(inst)
        if self.pool.get(path_str(inst.path)) is inst:
            self.retained_records += 1

    def activate(self, eng: Engine, value: Any) -> None:
        if not self.active or eng.no_reg > 0:
            return
        if isinstance(value, RecInst):
            if value.eng is not eng or value in self.live:
                return
            self.register(eng, value)
            for slot in value.slots.values():
                if slot.state == "ok":
                    self.activate(eng, slot.value)
        elif isinstance(value, ArrV):
            for v in value.items:
                self.activate(eng, v)
        elif isinstance(value, MapV):
            for v in value.entries.values():
                self.activate(eng, v)

    def compute(self, eng: Engine, key: str, run: Callable[[], Any]) -> Any:
        value = self.revisions.compute(eng, key, run)
        self.activate(eng, value)
        return value

    def absent(self, key: str) -> None:
        if self.active:
            self.revisions.accept(key, ABSENT)

    def prune(self, eng: Engine) -> None:
        # Captured scopes may refer back to their record. Explicitly prune the
        # metadata: a weak-key table's strong values would retain that cycle.
        live = set(eng.env.registry)
        self.inputs = {inst: raw for inst, raw in self.inputs.items() if inst in live}
        self.roots = {n: v for n, v in self.roots.items() if n in eng.env.roots}

    def finish(self, eng: Engine) -> None:
        self.prune(eng)
        if not self.active:
            return
        for key, (inst, _) in list(eng.slots_by_key.items()):
            if inst not in self.live:
                del eng.slots_by_key[key]
                eng.reads.pop(key, None)
        # Assertions run again after settling; removed roots and assertions
        # have no cached result whose old reads need to survive the edit.
        for key in list(eng.reads):
            if key.startswith("assert:") or (
                key.startswith("root:") and key[5:] not in eng.env.roots
            ):
                del eng.reads[key]
        self.revisions.finish()
        self.revisions.prune(eng)
        self.pool.clear()
        self.live.clear()
        self.invalid.clear()
        self.root_values.clear()
        self.root_reads.clear()
        self.active = False

File: decl-ts/src/qengine/edits.ts
// A Session revision replays root binding over retained record identities.
// The query graph verifies individual member inputs, values and membership.
import type { Engine, Scope } from '../engine.ts';
import type { RecInst, RT, Seg, Slot, Value } from '../semantics.ts';
import {
  ABSENT,
  isArr,
  isMap,
  isQ,
  isRange,
  isRec,
  isRef,
  pathStr,
  valueEq,
} from '../semantics.ts';
import { Revisions } from './revisions.ts';

function sameRaw(a: Value, b: Value): boolean {
  if (a === b) return true;
  if (Array.isArray(a) && Array.isArray(b))
    return a.length === b.length && a.every((x, i) => sameRaw(x, b[i]));
  if (a?.__jobj && b?.__jobj)
    return (
      a.entries.length === b.entries.length &&
      a.entries.every(
        ([k, v]: [string, Value], i: number) =>
          k === b.entries[i][0] && sameRaw(v, b.entries[i][1]),
      )
    );
  if (a?.__expr && b?.__expr) {
    const x: Scope = a.scope,
      y: Scope = b.scope;
    return (
      a.__expr === b.__expr &&
      x.inst === y.inst &&
      x.menv === y.menv &&
      x.rootName === y.rootName &&
      x.locals.size === y.locals.size &&
      [...x.locals].every(([k, v]) => y.locals.has(k) && sameRaw(v, y.locals.get(k)))
    );
  }
  if (isQ(a) && isQ(b)) return valueEq(a, b);
  return false;
}

/** Snapshot the observable shape before records receive their next slot maps.
 * A record's values are separate member queries; whole-value consumers record
 * an aggregate observation, which is conservatively invalidated on edits. */
function capture(v: Value): (next: Value) => boolean {
  if (isRec(v)) {
    const rt = v.rt,
      owner = (v as any).eng,
      path = pathStr(v.path),
      entries = [...v.entryOrder];
    const members = [...v.slots].map(([n, s]) => [n, s.kind, !!s.hidden]);
    const extras = [...v.extras].map(([k, x]) => [k, capture(x)] as const);
    return (n) =>
      isRec(n) &&
      n === v &&
      n.rt === rt &&
      (n as any).eng === owner &&
      pathStr(n.path) === path &&
      n.entryOrder.length === entries.length &&
      n.entryOrder.every((k: string, i: number) => k === entries[i]) &&
      n.slots.size === members.length &&
      [...n.slots].every(
        ([k, s], i) =>
          k === members[i][0] && s.kind === members[i][1] && !!s.hidden === members[i][2],
      ) &&
      n.extras.size === extras.length &&
      [...n.extras].every(([k, x], i) => k === extras[i][0] && extras[i][1](x));
  }
  if (isArr(v)) {
    const path = pathStr(v.path),
      items = v.items.map(capture);
    return (n) =>
      isArr(n) &&
      pathStr(n.path) === path &&
      n.items.length === items.length &&
      items.every((matches: (x: Value) => boolean, i: number) => matches(n.items[i]));
  }
  if (isMap(v)) {
    const path = pathStr(v.path),
      entries = [...v.entries].map(([k, x]) => [k, capture(x)] as const);
    return (n) =>
      isMap(n) &&
      pathStr(n.path) === path &&
      n.entries.size === entries.length &&
      [...n.entries].every(([k, x], i) => k === entries[i][0] && entries[i][1](x));
  }
  if (isRef(v)) {
    if (v.snap || v.inverseRef || v.roundRef) return () => false;
    return (n) => isRef(n) && !(n.snap || n.inverseRef || n.roundRef) && valueEq(v, n);
  }
  if (v === null || typeof v !== 'object' || isQ(v) || isRange(v)) return (n) => valueEq(v, n);
  return () => false;
}

export class Edits {
  active = false;
  readonly revisions = new Revisions();
  slotComputes = 0;
  retainedRecords = 0;
  preparedQueries = 0;
  private readonly inputs = new WeakMap<RecInst, Map<string, Value>>();
  private readonly roots = new Map<string, { raw: Value; expression: boolean }>();
  private pool = new Map<string, RecInst>();
  private live = new Set<RecInst>();
  private invalid = new Set<string>();
  private rootValues = new Map<string, Value>();
  private rootReads = new Map<string, Set<string>>();
  private bindingReuse = true;

  constructor() {
    this.revisions.resolve = (eng, key) => {
      const slot = eng.slotsByKey.get(key);
      if (slot) {
        eng.forceSlot(slot.inst, slot.name);
        return true;
      }
      if (key.startsWith('const:')) {
        const [index, name] = key.slice(6).split('|');
        const env = [...eng.constEnvs][Number(index)];
        if (env?.consts.has(name)) {
          eng.forceConstIn(env, name, '');
          return true;
        }
      }
      return false;
    };
  }

  begin(eng: Engine, changed: Set<string>, documents: Map<string, Value>): void {
    const values = new Map<string, Value>(),
      forced = new Set<string>();
    const errors = eng.env.diagnostics.length > 0;
    this.bindingReuse = !errors;
    this.pool = new Map(eng.env.registry.map((r) => [pathStr(r.path), r]));
    this.live.clear();
    this.retainedRecords = 0;
    this.rootValues = new Map(eng.env.roots);
    this.rootReads = new Map(
      [...eng.env.roots.keys()].map((n) => [`root:${n}`, new Set(eng.reads.get(`root:${n}`))]),
    );
    const knownInputs = new Set<string>();
    const collect = (v: Value, into: Set<string>): void => {
      if (isRec(v))
        for (const [n, slot] of v.slots) {
          const key = `${pathStr(v.path)}.${n}`;
          eng.slotsByKey.set(key, { inst: v, name: n });
          into.add(key);
          if (slot.state === 'ok') collect(slot.value, into);
        }
      else if (isArr(v)) v.items.forEach((x: Value) => collect(x, into));
      else if (isMap(v)) [...v.entries.values()].forEach((x) => collect(x, into));
    };
    const diff = (before: Value, next: Value, value: Value, producer: string): void => {
      if (sameRaw(before, next)) return;
      knownInputs.add(producer);
      forced.add(producer);
      if (isRec(value) && before?.__jobj && next?.__jobj) {
        const a = new Map<string, Value>(before.entries),
          b = new Map<string, Value>(next.entries);
        for (const name of new Set([...a.keys(), ...b.keys()])) {
          if (a.has(name) === b.has(name) && sameRaw(a.get(name), b.get(name))) continue;
          const slot = value.slots.get(name);
          if (slot) {
            const key = `${pathStr(value.path)}.${name}`;
            eng.slotsByKey.set(key, { inst: value, name });
            diff(a.get(name), b.get(name), slot.value, key);
            forced.add(key);
          }
        }
      } else if (isArr(value) && Array.isArray(before) && Array.isArray(next)) {
        for (let i = 0; i < Math.max(before.length, next.length); i++)
          diff(before[i], next[i], value.items[i], producer);
      } else if (isMap(value) && before?.__jobj && next?.__jobj) {
        const a = new Map<string, Value>(before.entries),
          b = new Map<string, Value>(next.entries);
        for (const k of new Set([...a.keys(), ...b.keys()]))
          diff(a.get(k), b.get(k), value.entries.get(k), producer);
      } else collect(value, forced);
    };
    for (const name of changed) {
      const previous = this.roots.get(name),
        value = eng.env.roots.get(name),
        key = `root:${name}`;
      forced.add(key);
      knownInputs.add(key);
      if (previous && !previous.expression && documents.has(name))
        diff(previous.raw, documents.get(name), value, key);
      else collect(value, forced);
    }
    const readers = new Map<string, string[]>();
    for (const [key, deps] of eng.reads)
      for (const dep of deps) {
        (readers.get(dep) ?? readers.set(dep, []).get(dep)!).push(key);
        if (['edge:', 'snapshot:', 'round:', 'value:'].some((p) => dep.startsWith(p)))
          forced.add(key);
      }
    for (const [name, value] of eng.env.roots) values.set(`root:${name}`, value);
    for (const inst of eng.env.registry)
      for (const [name, slot] of inst.slots) {
        const key = `${pathStr(inst.path)}.${name}`;
        if (slot.state === 'ok' || slot.state === 'absent')
          values.set(key, slot.state === 'absent' ? ABSENT : slot.value);
        if (errors) {
          forced.add(key);
          eng.slotsByKey.set(key, { inst, name });
        }
      }
    for (const [index, env] of [...eng.constEnvs].entries())
      for (const [name, con] of env.consts) {
        const key = `const:${index}|${name}`;
        if (con.state === 'ok') values.set(key, con.value);
        if (errors) forced.add(key);
      }
    // Captured scopes and reference answers belong to their original owner.
    // A matching path alone cannot make such a result reusable in a new edit.
    const retain = (v: Value, seen = new Set<Value>()): boolean => {
      if (v === null || typeof v !== 'object' || isQ(v) || isRange(v)) return true;
      if (seen.has(v)) return true;
      seen.add(v);
      if (isRef(v)) return !(v.snap || v.inverseRef || v.roundRef);
      if (isRec(v))
        return (
          (v as any).eng === eng &&
          [...(this.inputs.get(v)?.values() ?? [])].every((x) => retain(x, seen))
        );
      if (isArr(v)) return v.items.every((x: Value) => retain(x, seen));
      if (isMap(v)) return [...v.entries.values()].every((x) => retain(x, seen));
      if (Array.isArray(v)) return v.every((x) => retain(x, seen));
      if (v.__jobj) return v.entries.every(([, x]: [string, Value]) => retain(x, seen));
      if (v.__expr) {
        const sc: Scope = v.scope;
        return (
          (!sc.inst || (sc.inst as any).eng === eng) &&
          [...sc.locals.values()].every((x) => retain(x, seen))
        );
      }
      return false;
    };
    if (eng.queried.size) for (const [key, value] of values) if (!retain(value)) forced.add(key);
    if (errors) for (const key of eng.reads.keys()) forced.add(key);
    const aggregate = [...eng.reads.values()].some((deps) =>
      [...deps].some((d) => d.startsWith('value:')),
    );
    const invalid = new Set<string>(),
      queue = [...forced];
    if (aggregate) queue.push(...values.keys());
    while (queue.length) {
      const key = queue.pop()!;
      if (invalid.has(key)) continue;
      invalid.add(key);
      // An expression producer can change its descendants' captured inputs.
      // Direct document changes were already diffed down to supplied members.
      if (!knownInputs.has(key) && values.has(key)) {
        const descendants = new Set<string>();
        collect(values.get(key), descendants);
        for (const child of descendants) if (!invalid.has(child)) queue.push(child);
      }
      queue.push(...(readers.get(key) ?? []));
    }
    this.invalid = invalid;
    this.preparedQueries = invalid.size;
    this.revisions.beginQueries(eng, invalid, values, forced, capture);
    const copied = new Set<RecInst>();
    for (const key of invalid) {
      const found = eng.slotsByKey.get(key);
      if (!found) continue;
      const slot = found.inst.slots.get(found.name);
      if (!slot?.compute) continue;
      if (!copied.has(found.inst)) {
        found.inst.slots = new Map(found.inst.slots);
        copied.add(found.inst);
      }
      found.inst.slots.set(found.name, { ...slot, state: 'unforced', value: undefined });
    }
    for (const [index, env] of [...eng.constEnvs].entries())
      for (const [name, con] of env.consts)
        if (invalid.has(`const:${index}|${name}`)) {
          con.state = 'unforced';
          delete con.value;
        }
    eng.env.roots.clear();
    eng.env.registry.splice(0);
    eng.env.diagnostics.splice(0);
    eng.failedInputs.clear();
    eng.deferredSlots = [];
    eng.deferredRoots = [];
    eng.phase = 1;
    eng.settled = false;
    eng.prev = null;
    eng.snap = null;
    eng.queried.clear();
    eng.refIndex.clear();
    eng.computingEdges.clear();
    eng.edgeBases = [];
    eng.roundRoots = null;
    eng.snapshotValue = null;
    this.active = true;
  }

  root(eng: Engine, name: string, raw: Value, expression: boolean, run: () => Value): Value {
    const previous = this.roots.get(name);
    this.roots.set(name, { raw, expression });
    if (!this.active) return run();
    const key = `root:${name}`;
    if (
      !previous ||
      previous.expression !== expression ||
      !(expression ? previous.raw === raw : sameRaw(previous.raw, raw))
    )
      this.revisions.force(key);
    if (!this.invalid.has(key) && this.rootValues.has(name)) {
      eng.reads.set(key, this.rootReads.get(key) ?? new Set());
      const value = this.rootValues.get(name);
      this.activate(eng, value);
      return value;
    }
    return this.compute(eng, key, run);
  }

  unchanged(inst: RecInst, entries: [string, Value][]): boolean {
    const before = this.inputs.get(inst);
    return (
      this.bindingReuse &&
      !!before &&
      before.size === entries.length &&
      inst.entryOrder.length === entries.length &&
      entries.every(
        ([k, v], i) => inst.entryOrder[i] === k && before.has(k) && sameRaw(before.get(k), v),
      )
    );
  }

  record(rt: RT, path: Seg[], parent: RecInst | null): RecInst | undefined {
    if (!this.active) return undefined;
    const old = this.pool.get(pathStr(path));
    return old && old.rt === rt && old.parent === parent ? old : undefined;
  }

  bound(
    eng: Engine,
    inst: RecInst,
    entries: [string, Value][],
    oldSlots?: Map<string, Slot>,
  ): void {
    const before = this.inputs.get(inst),
      next = new Map(entries);
    if (this.active)
      for (const [name, slot] of inst.slots) {
        const key = `${pathStr(inst.path)}.${name}`;
        if (
          !before ||
          before.has(name) !== next.has(name) ||
          !sameRaw(before.get(name), next.get(name)) ||
          oldSlots?.get(name)?.kind !== slot.kind
        )
          this.revisions.force(key);
        else if (oldSlots?.has(name)) inst.slots.set(name, oldSlots.get(name)!);
        eng.slotsByKey.set(key, { inst, name });
      }
    this.inputs.set(inst, next);
    if (this.active)
      for (const slot of inst.slots.values())
        if (slot.state === 'ok') this.activate(eng, slot.value);
  }

  register(eng: Engine, inst: RecInst): void {
    if (this.live.has(inst)) return;
    this.live.add(inst);
    eng.env.registry.push(inst);
    if (this.pool.get(pathStr(inst.path)) === inst) this.retainedRecords++;
  }

  activate(eng: Engine, value: Value): void {
    if (!this.active || eng.noReg > 0) return;
    if (isRec(value)) {
      if ((value as any).eng !== eng || this.live.has(value)) return;
      this.register(eng, value);
      for (const s of value.slots.values()) if (s.state === 'ok') this.activate(eng, s.value);
    } else if (isArr(value)) for (const v of value.items) this.activate(eng, v);
    else if (isMap(value)) for (const v of value.entries.values()) this.activate(eng, v);
  }

  compute(eng: Engine, key: string, run: () => Value): Value {
    const value = this.revisions.compute(eng, key, run);
    this.activate(eng, value);
    return value;
  }

  absent(key: string): void {
    if (this.active) this.revisions.accept(key, ABSENT);
  }

  finish(eng: Engine): void {
    if (!this.active) return;
    for (const [key, found] of eng.slotsByKey)
      if (!this.live.has(found.inst)) {
        eng.slotsByKey.delete(key);
        eng.reads.delete(key);
      }
    for (const name of this.roots.keys()) if (!eng.env.roots.has(name)) this.roots.delete(name);
    // Assertions run again after settling, so their old reads have no cache
    // to preserve. Drop removed roots along with removed member queries.
    for (const key of eng.reads.keys())
      if (
        key.startsWith('assert:') ||
        (key.startsWith('root:') && !eng.env.roots.has(key.slice(5)))
      )
        eng.reads.delete(key);
    this.revisions.finish();
    this.revisions.prune(eng);
    this.pool.clear();
    this.live.clear();
    this.rootValues.clear();
    this.rootReads.clear();
    this.invalid.clear();
    this.active = false;
  }
}
